src/main/java/org/terasology/alterationEffects/speed/WalkSpeedAlterationEffect.java

```diff
@@ -15,112 +15,32 @@
-                walkSpeed.multiplier = modifiedMagnitude;
-                modifiersFound = true;
-            }
-        }
-
-        // If the modified duration is between the accepted values (0 and Long.MAX_VALUE), and the base duration is not infinite,
-        // add a delayed action to the DelayManager using the new system.
-        if (modifiedDuration < Long.MAX_VALUE && modifiedDuration > 0 && duration != AlterationEffects.DURATION_INDEFINITE) {
-            String effectID = effectModifyEvent.getEffectIDWithShortestDuration();
-            delayManager.addDelayedAction(entity, AlterationEffects.EXPIRE_TRIGGER_PREFIX + AlterationEffects.WALK_SPEED + "|" + effectID, modifiedDuration);
-        }
-        // Otherwise, if the duration is greater than 0, there are no modifiers found, and the effect modify event was not consumed,
-        // add a delayed action to the DelayManager using the old system.
-        else if (duration > 0 && !modifiersFound && !effectModifyEvent.isConsumed()) {
-            delayManager.addDelayedAction(entity, AlterationEffects.EXPIRE_TRIGGER_PREFIX + AlterationEffects.WALK_SPEED, duration);
-        }
-        // Otherwise, if there are either no modifiers found, or none of the modifiers collected in the event have infinite
-        // duration, remove the component associated with this walk speed effect.
-        else if (!modifiersFound || !effectModifyEvent.getHasInfDuration()) {
-            entity.removeComponent(WalkSpeedComponent.class);
-        }
-        // If this point is reached and none of the above if-clauses were met, that means there was at least one modifier
-        // collected in the event which has infinite duration.
+    protected WalkSpeedComponent upsertComponent(Optional<WalkSpeedComponent> maybeComponent, float magnitude,
+                                                 long duration) {
+        WalkSpeedComponent walkSpeed = maybeComponent.orElse(new WalkSpeedComponent());
+        walkSpeed.multiplier = magnitude;
+        return walkSpeed;
     }
 
-    /**
-     * This will apply the walk speed effect on the given entity by calling the method
-     * {@link #applyEffect(EntityRef, EntityRef, float, long)}.
-     *
-     * @param instigator    The entity who applied the walk speed effect.
-     * @param entity        The entity that the walk speed effect is being applied on.
-     * @param id            Inapplicable to the walk speed effect.
-     * @param magnitude     The magnitude of the walk speed effect.
-     * @param duration      The duration of the walk speed effect.
-     */
-    @Override
-    public void applyEffect(EntityRef instigator, EntityRef entity, String id, float magnitude, long duration) {
-        applyEffect(instigator, entity, magnitude, duration);
+    protected WalkSpeedComponent updateComponent(OnEffectModifyEvent event, WalkSpeedComponent component) {
+        component.multiplier = event.getMagnitudeResultValue();
```

Comment:
So first I have a general doubt
-Do I get this correctly? upsertComponent() would be used at the time of initial application of effect and updateComponent() would be called to react to the OnEffectModifyEvent to add modifiers ?

Next this is related to the implementation of the alteration effects even before this PR..
> component.multiplier = event.getMagnitudeResultValue();

Here we are changing the multiplier of the component to the result value provided by the OnEffectModifyEvent. 
But, the OnEffectModifyEvent, even though uses the base magnitude to calculate the result value, the baseMagnitude we are sending is zero.
>OnEffectModifyEvent effectModifyEvent = entity.send(
                new OnEffectModifyEvent(instigator, entity, 0, 0, this, id)
        );

Which would mean event.getMagnitudeResultValue() only returns a value consisting of modifiers which are supposed to modify the original value and we completely overlook the original magnitude which was added by upsertComponent() so the value we receive wouldn't be the intended one.

I think the use of alteration effects got jumbled up a bit...., I think the OnEffectModifyEvent was supposed to be a version of AbstractValueModifiableEvent... 
It was supposed to send a base value and other modules were supposed to add modifiers and multipliers...
Now since we send a 0 base magnitude the modifiers now have to use the addMagnitude(value) method in OnEffectModifyEvent (which adds a modifier) to reapply the same value!
What we wanted -> (baseMagnitude + modifier)
what we end up doing -> (0 + modifier) 
and then the method has to ensure that modifier sent is equal to baseMagnitude to get the Effect working with a base magnitude and not a modified one..
I think you might not have completely understood what I'm trying to say.... please ask for clarification... I really really want to fix this.

Reply:
> Do I get this correctly? upsertComponent() would be used at the time of initial application of effect and updateComponent() would be called to react to the OnEffectModifyEvent to add modifiers ?

Yes, that's correct. I used names describing what is happening there, but we could also use names that convey the semantics of what should happen.

> Which would mean event.getMagnitudeResultValue() only returns a value consisting of modifiers which are supposed to modify the original value and we completely overlook the original magnitude which was added by upsertComponent() so the value we receive wouldn't be the intended one.

That's indeed right - I think one of the insights @jdrueckert had the last time we looked at this was that those values can either be removed _or_ should be send out with the modify event. I did not want to change anything regarding the logic in this PR, but that's exactly the part I hoped would become clearer when looking at it in isolation.

> This means, if we do send out the correct baseMagnitude then the following statement in Equipments will no longer be needed

Yep, that's exactly what we currently need to do
- _apply_ an alteration effect giving it `magnitude` and `duration` 
- write a system to _react_ upon the modify event to set the exact same `magnitude`

With these changes merged it would become way easier to change this behavior and test in-game whether everything still works as expected. I assume some things will just start working, while others will be quirky with doubled strenght... We'll see ^^